### PR TITLE
Adding SGX basic validation test

### PR DESCRIPTION
**Pull request by shruti25ratnam**
Adding a testcase that verifies if the VM is SGX Enabled by running HelloWorld and Remote Attestation. Priority set as 1.

        Steps:
            1. Add keys and tool chain from Intel-SGX, LLVM and Microsoft repositories.
            2. Install DCAP driver if missing.
            3. Install required package.
            3. Run Helloworld and Remote Attestation tests.

The shell script that prepares an ACC VM and runs the tests (This takes 3.5 mins on DC2sv3):

```
_echo 'deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu focal main' | sudo tee /etc/apt/sources.list.d/intel-sgx.list
wget -qO - https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key | sudo apt-key add -

echo "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-10 main" | sudo tee /etc/apt/sources.list.d/llvm-toolchain-focal-10.list
wget -qO - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -

echo "deb [arch=amd64] https://packages.microsoft.com/ubuntu/20.04/prod focal main" | sudo tee /etc/apt/sources.list.d/msprod.list
wget -qO - https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -

sudo apt update
dmesg | grep -i sgx

sudo apt -y install clang-10 libssl-dev gdb libsgx-enclave-common libsgx-quote-ex libprotobuf17 libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client open-enclave

cp -r /opt/openenclave/share/openenclave/samples ~/mysamples

source /opt/openenclave/share/openenclave/openenclaverc

cd mysamples/helloworld/
make build
make run
cd ../attestation/
make build
make run_
```

### CODE REVIEW

File: microsoft/testsuites/acc/bvt.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+            3. Run Helloworld and Remote Attestation tests.
+
+        """,
+        priority=1,
+    )
+
+    def verify_SGX(self, log: Logger, node: Node) -> None:
+
+        if self.node.os.information.version != "20.4.0":
+                raise UnsupportedDistroException(self.node.os)
```
Comment (LiliDeng) [inline microsoft/testsuites/acc/bvt.py]: is it only support on ubuntu 20.04? should we check the os type? 
```
            if (
                isinstance(self.node.os, Ubuntu)
                and self.node.os.information.version != "20.4.0"
            ):
```

File: microsoft/testsuites/acc/bvt.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                ],
+        )
+
+        #Get apt dependencies from Intel-SGX repo
+        #
+        #<  echo 'deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu focal main' | 
+        #   sudo tee /etc/apt/sources.list.d/intel-sgx.list >
+
+        echo_command = 'echo \'deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu focal main\' | sudo tee /etc/apt/sources.list.d/intel-sgx.list'        
+        stdout = node.execute(echo_command,sudo=True,expected_exit_code=0,expected_exit_code_failure_message="failed to add apt key")
```
Comment (LiliDeng) [inline microsoft/testsuites/acc/bvt.py]: call node.tools[Echo].write_to_file

File: microsoft/testsuites/acc/bvt.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+        #Copy LLVM toolchain
+        #
+        #<  echo "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-10 main" | 
+        #   sudo tee /etc/apt/sources.list.d/llvm-toolchain-focal-10.list >
+
+        deb_command = '"deb http://apt.llvm.org/focal/ llvm-toolchain-focal-10 main" | sudo tee /etc/apt/sources.list.d/llvm-toolchain-focal-10.list'
+        node.execute( cmd=f'deb_command',expected_exit_code=0,expected_exit_code_failure_message="failed to fetch llvm toolchain")
```
Comment (LiliDeng) [inline microsoft/testsuites/acc/bvt.py]: same as above

File: microsoft/testsuites/acc/bvt.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                    "https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key",
+                ],
+        )
+
+        #Get apt dependencies from Intel-SGX repo
+        #
+        #<  echo 'deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu focal main' | 
+        #   sudo tee /etc/apt/sources.list.d/intel-sgx.list >
+
+        echo_command = 'echo \'deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu focal main\' | sudo tee /etc/apt/sources.list.d/intel-sgx.list'
```
Comment (LiliDeng) [inline microsoft/testsuites/acc/bvt.py]: replace focal into self.node.os.information.codename?

File: microsoft/testsuites/acc/bvt.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                    keys_location=["https://apt.llvm.org/llvm-snapshot.gpg.key"],
+        )
+        
+        #Get Ubuntu 20.04 packages from Microsoft prod repo
+        #
+        # < echo "deb [arch=amd64] https://packages.microsoft.com/ubuntu/20.04/prod focal main" | 
+        #   sudo tee /etc/apt/sources.list.d/msprod.list >
+
+        echo_command = 'echo "deb [arch=amd64] https://packages.microsoft.com/ubuntu/20.04/prod focal main" | sudo tee /etc/apt/sources.list.d/msprod.list'
+        stdout = node.execute(echo_command,sudo=True,expected_exit_code=0,expected_exit_code_failure_message="failed to get packages")
```
Comment (LiliDeng) [inline microsoft/testsuites/acc/bvt.py]: same as above

File: microsoft/testsuites/acc/bvt.py (hunk 2 above)
Comment (shruti25ratnam) [reply microsoft/testsuites/acc/bvt.py]: Sure I'll use that instead. Thanks!

File: microsoft/testsuites/acc/bvt.py (hunk 1 above)
Comment (LiliDeng) [reply microsoft/testsuites/acc/bvt.py]: UnsupportedDistroException will cause case fail, skipexception can be raised here.

Comment (shruti25ratnam) [reply microsoft/testsuites/acc/bvt.py]: Yes, currently this case only supports one Ubuntu version. I changed it to 18.04 for ease. Sure, I will use skip Exception instead.

**Comment by dsrivastavv** [comment]
@shruti25ratnam The changes enabling ACC features are merged. Can you please update this PR accordingly?

File: microsoft/testsuites/acc/bvt.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+from lisa.operating_system import Debian
+from lisa.tools import Dmesg, Make, Wget, Whoami
+from lisa.util import SkippedException
+
+
+@TestSuiteMetadata(
+    area="ACC_BVT",
+    category="functional",
+    description="""
+    This Basic Validation Test (BVT) suite validates the availability of Secure Guard \
```
Comment (dsrivastavv) [inline microsoft/testsuites/acc/bvt.py]: [nit] `\` is not needed

File: microsoft/testsuites/acc/bvt.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        ),
+    )
+    def verify_sgx(self, log: Logger, node: Node) -> None:
+
+        if node.os.information.version == "18.4.0":
+            os_version = "18.04"
+        elif node.os.information.version == "20.4.0":
+            os_version = "20.04"
+        else:
+            raise SkippedException(node.os)
```
Comment (dsrivastavv) [inline microsoft/testsuites/acc/bvt.py]: Can you please add complete message. Something like "os version: {} is not supported"

File: microsoft/testsuites/acc/bvt.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+        """,
+        priority=1,
+        requirement=simple_requirement(
+            supported_features=[acc.ACC],
+        ),
+    )
+    def verify_sgx(self, log: Logger, node: Node) -> None:
+
+        if node.os.information.version == "18.4.0":
```
Comment (dsrivastavv) [inline microsoft/testsuites/acc/bvt.py]: You would also need to verify that the OS is ubuntu. It's possible multiple distros have this version

File: microsoft/testsuites/acc/bvt.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        node.os.install_packages(
+            f"clang-10 libssl-dev gdb libsgx-enclave-common libsgx-quote-ex "
+            f"{libprotobuf} libsgx-dcap-ql libsgx-dcap-ql-dev "
+            "az-dcap-client open-enclave"
+        )
+        # REMOVE install_cmd = "apt -y install clang-10 libssl-dev gdb
+        # libsgx-enclave-common libsgx-quote-ex libprotobuf10
+        # libsgx-dcap-ql libsgx-dcap-ql-dev az-dcap-client
+        #  open-enclave "
+        # stdout = node.execute(install_cmd, sudo=True)
```
Comment (dsrivastavv) [inline microsoft/testsuites/acc/bvt.py]: [nit] remove debugging comments

File: microsoft/testsuites/acc/bvt.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        )
+        node.execute(copy_cmd, shell=True)
+
+        # Run Hello World and Remote Attestation
+        helloworld_dir = samples_folder / "helloworld"
+        attestation_dir = samples_folder / "attestation"
+        source_command = ". /opt/openenclave/share/openenclave/openenclaverc"
+        fail_msg = "HELLO WORLD TEST FAILED"
+
+        node.tools[Make]
```
Comment (dsrivastavv) [inline microsoft/testsuites/acc/bvt.py:162]: Why is this required? If `make` is a prerequisite, you can directly call install. @squirrelsc Is there a way to add a tool as a dependency?

File: microsoft/testsuites/acc/bvt.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+        assert_that(result.stdout).described_as("error message").contains(
+            "Enclave called into host to print: Hello World!"
+        )
+
+        fail_msg = "REMOTE ATTESTATION HAS FAILED"
+        username = node.tools[Whoami].get_username()
+        node.execute(f"usermod -a -G sgx_prv {username}", shell=True, sudo=True)
+        node.execute("export SGX_AESM_ADDR=1", shell=True, sudo=True)
+        result = node.execute(
```
Comment (dsrivastavv) [inline microsoft/testsuites/acc/bvt.py:179]: Can you please add a new tool for `sgx` and move the installation there? This will make it easy to share in other test cases. Please feel free to ping me if you need help writing a tool in Lisa.

File: microsoft/testsuites/acc/bvt.py (hunk 10 above)
Comment (shruti25ratnam) [reply microsoft/testsuites/acc/bvt.py:162]: make was actually supposed to be pre-installed on the platform, it somehow needed re-instantiating during our test runs. So, we kept node.tools[Make].

File: microsoft/testsuites/acc/bvt.py (hunk 8 above)
Comment (shruti25ratnam) [reply microsoft/testsuites/acc/bvt.py]: there's an assert (node.os,Debian) after this that will prevent non-Ubuntu package. I will move that above the check, that will make more sense, as you pointed out.

File: microsoft/testsuites/acc/bvt.py (hunk 11 above)
Comment (shruti25ratnam) [reply microsoft/testsuites/acc/bvt.py:179]: I am creating a separate work item for this on LISA repo for me. So, resolving this for now.

File: microsoft/testsuites/acc/bvt.py [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+        """,
+        priority=1,
+        requirement=simple_requirement(
+            supported_features=[acc.ACC],
+        ),
+    )
+    def verify_sgx(self, log: Logger, node: Node) -> None:
+
+        assert isinstance(node.os, Debian)
```
Comment (squirrelsc) [inline microsoft/testsuites/acc/bvt.py]: The assertion will make other distro failed on this test case. If they are not supported, it should raise a skipped exception.
If it needs to be asserted, it's better to provide an error message.

**Review by dsrivastavv — DISMISSED** [review]
Looks good now. Can you please clear your commit history before merging since we usually rebase and merge PR's?

**Comment by squirrelsc** [comment]
> Looks good now. Can you please clear your commit history before merging since we usually rebase and merge PR's?

Yes, we encourage people to organize meaningful commits, and minimize changes in each PR. For this one, I can squash the commits into one, when merging, because it's not big.

**Review by dsrivastavv — DISMISSED** [review]

**Review by squirrelsc — APPROVED** [review]
